pkg/cloud-init/cloud-init.go

```diff
@@ -120,6 +134,36 @@ func cloudInitUUIDFromVMI(vmi *v1.VirtualMachineInstance) string {
 	return string(vmi.Spec.Domain.Firmware.UUID)
 }
 
+func generateHostname(vmi *v1.VirtualMachineInstance, userData string) (string, error) {
+	const errPrefix = "cannot generate hostname:"
+
+	// This indicates that userdata is of YAML format.
+	// For more info: https://cloudinit.readthedocs.io/en/latest/topics/format.html
+	if !strings.HasPrefix(userData, "#cloud-config") && !strings.HasPrefix(userData, "Content-Type: text/cloud-config") {
+		if vmi == nil {
+			return "", fmt.Errorf("%s hostname was not defined in user data and vmi is nil", errPrefix)
+		}
+
+		return dns.SanitizeHostname(vmi), nil
+	}
+
+	// Remove comments
+	commentRegex := regexp.MustCompile("(?m)^#.*")
+	userData = commentRegex.ReplaceAllString(userData, "")
+
+	u := UserData{}
+	err := yaml.Unmarshal([]byte(userData), &u)
+	if err != nil {
+		return "", fmt.Errorf("%s error unmarshalling cloud init user data: %v", errPrefix, err)
+	}
+
+	if u.Hostname != "" {
+		return u.Hostname, nil
+	}
```

Comment:
So with this change, `Hostname` from userdata will take precedence over the `vmi.Spec.Hostname` field, correct? I have a couple of questions in this regard:
- Is that what we want to have? Shouldn't it be the opposite that the vmi field has a higher prio to make it backward-compatible (since previously the hostname from cloud-init was a no-op?)?
- Current implementation takes `vmi.Spec.Hostname` and if it's not empty, assignes it to the `pod.Spec.Hostname`: https://github.com/kubevirt/kubevirt/blob/307cd6a6921de9ddf35b92059d9c6f2bcb3eb97c/pkg/virt-controller/services/template.go#L448-L461 Do we need to consider the hostname provided via cloud-init for the pod as well?

Reply:
Thanks @vasiliy-ul for your great insights.

Regarding `vmi.Spec.Hostname` having precedence - you're absolutely right. Thanks for pointing that out. I've changed the logic (in https://github.com/kubevirt/kubevirt/pull/8483/commits/52df61f76b968a646308f3c2586e82f333c5a5af) so that if a hostname is defined in the vmi's spec - that's the hostname that is returned from `generateHostname()` function. A unit test is also added (in https://github.com/kubevirt/kubevirt/pull/8483/commits/976af1e868c1bceaa9bc487ee7bc6cc826a948bc).

Regarding your second point - I find it pretty weird that changing the vmi's hostname actually changes the pod's hostname. These specs should configure the **guest's** hostname, I'm not sure why changing the pod's hostname makes sense. In any case - the right and recommended way to set a custom hostname is through `vmi.Spec.Hostname`. Using cloud-init is a way to bypass our specs and while we want to support it, I think it's ok that the two would not cause the exact same affect. After all, cloud-init can also provide a bash script as a userData, and if that's the case we won't support it whatsoever. Maybe we should even pop up a warning that suggests to set `vmi.Spec.Hostname` instead. Does this make sense?

Reply:
> Regarding your second point - I find it pretty weird that changing the vmi's hostname actually changes the pod's hostname.

Yeah, my only concern here is if we rely somehow on that behavior and expect that the hostname of the pod is always aligned with vmi. But probably not... or at least we shouldn't, I think.

> Using cloud-init is a way to bypass our specs and while we want to support it, I think it's ok that the two would not cause the exact same affect. After all, cloud-init can also provide a bash script as a userData, and if that's the case we won't support it whatsoever.

Good point about a bash script. Makes sense.